Issue92without model development

aanersc opened the pull request:
viewer is developed without tests on 1e57839.

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -24,30 +23,18 @@ def instantiation_and_set_form_defaults(self):
-                self.forms[counter] = element_form_class({})
-
+            self.forms[counter] = element_form_class({})  # instantiation
+            self.forms[counter].set_defaults()  # setting the defaults
```
If SearchForm.__init__() calls set_default(). This is not needed.
```suggestion
```

aanersc replied:
according to refusal of above suggestion, this is also denied consequently.

aanersc replied:
It needed some converting of data of POST which makes code more complicated than the current one

aperrin66 replied:
The call to this method can be avoided if you use a standard `Form` with dynamic initial values, as suggested in an other comment

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -24,30 +23,18 @@ def instantiation_and_set_form_defaults(self):
     def get(self, request, *args, **kwargs):
         ''' Render page if no data is given '''
-        # set default values of form
         self.instantiation_and_set_form_defaults()
```
Can you rename?
```suggestion
        self.create_forms()
```

aanersc replied:
done

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -24,30 +23,18 @@ def instantiation_and_set_form_defaults(self):
         ''' Render page if no data is given '''
-        # set default values of form
         self.instantiation_and_set_form_defaults()
         self.validation_and_error_check()
```
Can you rename?
```suggestion
        self.validate_forms()
```

aanersc replied:
done

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -56,55 +43,42 @@ def post(self, request, *args, **kwargs):
-        self.filtering_func()
-        return self.final_rendering(request)
-
+    def filtering_func(self, request):
```
Why do you need "func" in function name!? Of course it is a "function", what else?
```suggestion
    def filter_datasets(self, request):
```

aanersc replied:
Done

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -56,55 +43,42 @@ def post(self, request, *args, **kwargs):
+                # using the filter function of each form sequentially
+                ds = self.forms[counter].filter(ds)
+            self.ds = ds
+        elif (request.method == 'GET'):
```
Why you don't want filtering in the GET request? With the default values, it will anyway return all datasets. Either let both get() and post() use exactly the same function without if. Or make a separate function for get() and for post(). Also, don't assign to self.ds, better to return ds as output - it is more clean.
"Explicit is better than implicit"

```python
    def get(self):
        ...
        ds = self.get_all_datasets()
        ...

    def post(self):
        ....
        ds = self.get_all_datasets()
        ds = self.filter_datasets(ds)
        ...

    def get_all_datasets(self):
        return CatalogDataset.objects.all()

    def filter_datasets(self, ds):
        for form in self.forms:
            ds = form.filter(ds)
        return ds
```

aanersc replied:
The reason for bypassing the filtering action in the GET is that there is no need to do it. IF you execute a procedure which is not needed, it only slows your performance. So, there is no need to make the pipeline elongated for GET request. This is my reasoning as far as I know.

Please consider that we can only GET and POST request for this viewer, then this if statement for distinguishing between these two is enough and simple(not complicated). There is no need to take care of PUT request or DELETE ones. So this function with this single if can manage everything for us.





If we use `ds`, that `ds` is only works for that specific function. There we need to address them to be in the `self.ds` for the functions of the calss (here `set_dataset_context`) to use the result of previous functions in the workflow of this **class-based** view. The code won't work if we don't assign it to the `self.ds`. I have tested it based on your suggestion and it won't work.

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -56,55 +43,42 @@ def post(self, request, *args, **kwargs):
+
+        # modify attributes based on the forms cleaned data
+        self.filtering_func(request)
+        return self.final_rendering(request)
```
final_rendering should not call set_context which calls set_dataset_context. "Flat is better than nested"
Post should call all:
```suggestion
        ds = self.get_all_datasets()
        ds = self.filter_datasets(ds)
        contexts = self.set_context(ds)
        return render(request, self.main_template, context)
        
```

akorosov replied:
The same for get()

aanersc replied:
done

aanersc replied:
for both get and post functions

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -56,55 +43,42 @@ def post(self, request, *args, **kwargs):
 
     def final_rendering(self, request):
         ''' Render page based on several forms of data as well as some other context '''
+        self.context = {}
```
self.context is not needed. It is better to receive it from self.set_context()
```suggestion
        context = self.set_context()
        eturn render(request, self.main_template, context)
```

aanersc replied:
I have written the `set_context` function and `set_dataset_context` without return statement for easily overriding in the future. This is the reason for this style of coding (to the best of my knowledge). Does `writing the functions of the class with RETURN statement` allow us to override it in the future for ADASviewer without using append in the child class? I don't think so. 

Therefore, it is better to write them in this way (consequently save the results in the self.context[....]) and in the future the child class will be written like this:
```python
def set_dataset_context(self):
        super().set_dataset_context()
        setself.context['additional_context'] = self.blablabla
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -56,55 +43,42 @@ def post(self, request, *args, **kwargs):
             form_list[counter] = value
         # passing the form_list into context
         self.context['form_list'] = form_list
+        self.set_dataset_context()
+
+    def set_dataset_context(self): # excluded from other context for pagination development of django
```
Why do you need a separate function for just one line? 
```suggestion
```

aanersc replied:
Please have a look at my comments as well!! I wrote that in the comment:

# excluded from other context for pagination development of django


In the possible child class we need to override this function if we want to take advantage of django pagination in the future.

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -34,6 +40,18 @@ class Meta:
+        received_polygon= self.cleaned_data['polygon']
+        if received_polygon is not None:
+            ds = ds.filter(geographic_location__geometry__intersects= received_polygon)
+            return ds
+        else:
+            # in the case that user have no specified polygon
+            # there is no need to filter
+            return ds
```
It won't work in that way, I have tested it before the push. Therefore this suggestion is denied.

aanersc replied:
It won't in this way, I have tested it before push

aanersc commented on geospaas/base_viewer/views.py:
```diff
@@ -24,30 +23,18 @@ def instantiation_and_set_form_defaults(self):
-            else:  # instantiating without default values
-                self.forms[counter] = element_form_class({})
-
+            self.forms[counter] = element_form_class({})  # instantiation
```
Lets write it in this way because we have separated two action in two separate lines of 26 and 27 from each other. Thus for more clear code and showing the sequence of these actions, let it be as a for loop, please.

aanersc replied:
Done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -8,19 +10,23 @@
 
-    def __init__(self, *args, **kwargs):
-        #self.geographic_location = CatalogGeographicLocation
-        initial = kwargs.pop('initial', {})
```
done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -34,6 +40,18 @@ class Meta:
+    def filter(self,ds):
+        received_polygon= self.cleaned_data['polygon']
+        if received_polygon is not None:
+            ds = ds.filter(geographic_location__geometry__intersects= received_polygon)
```
done

aanersc replied:
done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -34,6 +40,18 @@ class Meta:
+        pass
+
+    def filter(self,ds):
+        received_polygon= self.cleaned_data['polygon']
```
done

aanersc replied:
done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -8,19 +10,23 @@
+        t0 = self.cleaned_data['time_coverage_start']
+        t1 = self.cleaned_data['time_coverage_end'] + timezone.timedelta(hours=24)
+        ds = ds.filter(Q(time_coverage_end__lt=t1) & Q(time_coverage_start__gt=t0))
+        src=[]
+        src.append(self.cleaned_data['source'])
```
done

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+from django.db.models import Q
+
+from leaflet.forms.widgets import LeafletWidget
+from geospaas.catalog.models import Dataset as CatalogDataset
```
If you don't have another `Dataset` class in this file, you don't need to give it an alias

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+
+from geospaas.base_viewer.models import Search_dummy_leaflet
+#from geospaas.base_viewer.models import Search_with_para
+class SearchFormBelow(forms.ModelForm):
```
If I didn't miss anything, you don't use any functionality of `forms.ModelForm` and could just use `forms.Form`

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+        model = CatalogDataset
+        fields = ['time_coverage_start','time_coverage_end','source']
+
+    def set_defaults(self):
```
Why not just use the dynamic initial value functionality?
https://docs.djangoproject.com/en/3.0/ref/forms/api/#dynamic-initial-values

aanersc replied:
Done

aperrin66 commented on geospaas/base_viewer/models.py:
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+from django.contrib.gis.db import models as geomodels
+
+class Search_dummy_leaflet(geomodels.Model):
```
classes should be named in CamelCase

aanersc replied:
Done.

aanersc replied:
Done

aperrin66 commented on geospaas/base_viewer/templates/base_viewer/origin_of_templates.html:
```diff
@@ -0,0 +1,71 @@
+			height: 100%;
+		}
+	</style>-->
+	<style>
```
Maybe move the CSS in a separate file in a `static` folder?

aanersc replied:
Done

aperrin66 commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,84 @@
+from django.shortcuts import render
+from django.db.models import Q
```
This import is not used

aanersc replied:
Done

aanersc replied:
Done

aanersc commented:
@akorosov 
Please review the my comments as well as reviewing the code in the cb7fc36.

aanersc commented on geospaas/base_viewer/admin.py:
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
+# Register your models here.
```
Done

aanersc replied:
Done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+
+from leaflet.forms.widgets import LeafletWidget
+from geospaas.catalog.models import Dataset as CatalogDataset
+from django.contrib.gis.db import models as geomodels
```
Done

aanersc replied:
Done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+
+from geospaas.base_viewer.models import Search_dummy_leaflet
+#from geospaas.base_viewer.models import Search_with_para
+class SearchFormBelow(forms.ModelForm):
```
Done

aanersc replied:
Done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+        ds = ds.filter(source__in=src)
+        return ds
+
+class SearchFormAbove(forms.ModelForm):
```
Done

aanersc replied:
Done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,60 @@
+        fields = ['time_coverage_start','time_coverage_end','source']
+
+    def set_defaults(self):
+        self.data['time_coverage_end'] = timezone.now().date()
```
Done. Thanks for your scrutinizing comment.

aanersc replied:
Done

aperrin66 commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,74 @@
+
+class IndexView(View):
+
+    form_class = [SpatialSearchForm, TimeAndSourceForm, ]
```
you might want to name this `form_classes` since it contains a list and not a single class

aanersc replied:
Django won't work with any other names. It must be `form_class`

aanersc commented:
@akorosov 
The third comment of friday short talk in applied on 7753537. The first and the second are still remaining ones and not implemented yet.

akorosov commented:
@aanersc , great!
Just one question from my curiosity. Why do you systematically prefer:
```python
        for counter in range(len(self.form_class)):
            ds = self.forms[counter].filter(ds)
```
to a more elegant, in my opinion, solution:
```python
        for form in self.forms:
            ds = form.filter(ds)
```

aanersc commented:
> @aanersc , great!
> Just one question from my curiosity. Why do you systematically prefer:
> 
> ```python
>         for counter in range(len(self.form_class)):
>             ds = self.forms[counter].filter(ds)
> ```
> 
> to a more elegant, in my opinion, solution:
> 
> ```python
>         for form in self.forms:
>             ds = form.filter(ds)
> ```

This is not my systematic preference. It is intentionally done with this approach for this case only. With the simple for loop, the filtering is accomplished by thier placement in the GUI. But, this approach opens our hands in the future to filter not only by the order of showing the forms in GUI, but also by the reverse order of it. **Maybe** in the future the reverse order of filtering helps us to filter faster the datasets based on reversing the filtering action or not.

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,91 @@
+    def create_forms(self, *args):#request_method='GET'
+        ''' Set default values for the form by instantiating them '''
+        self.forms = [i(*args) for i in self.form_class]
+        #if request_post is 'GET':
+        #    request_post = dict()
+        #if len(args)==0:
+        #     # GET method
+        #else:
+        #    self.forms = [i(args[0]) for i in self.form_class] # POST method
+        #return self.forms
+
+        #### initialize the form list for setting default values
+        #### loop for instantiating with defaults
+        ###self.forms = [i({}) for i in self.form_class]
+        ###for counter, element_form_class in enumerate(self.form_class):
+        ###    self.forms[counter].set_defaults()  # setting the defaults
```
```suggestion
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,91 @@
+
+    def get_filtered_datasets(self):
+        ds = self.get_all_datasets()
+        for counter in range(len(self.form_class)):
+            # using the filter function of each form sequentially
+            ds = self.forms[counter].filter(ds)
```
```suggestion
        for form in self.form_class:
            # using the filter function of each form sequentially
            ds = form.filter(ds)
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,91 @@
+    def set_context(self):
+        self.context = {}
+        # initializing the list of the forms that passed into context
+        form_list = [None] * len(self.forms)#[i for i in self.forms]
+        for i in range(len(self.forms)):
+            if isinstance(self.forms[i].__class__(), SpatialSearchForm):
+                form_list[0] = self.forms[i] # DICTATE the spatial search form to be the FIRST one
+            if isinstance(self.forms[i].__class__(), TimeAndSourceForm):
+                form_list[1] = self.forms[i] # DICTATE the time-and-date search form to be the SECOND one
+        # passing the form_list into context
+        self.context['form_list'] = form_list
```
```suggestion
        # passing the form_list into context
        self.context['form_list'] = self.forms
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,91 @@
+    viewname = 'index'
+    forms = list()
+
+    def create_forms(self, *args):#request_method='GET'
+        ''' Set default values for the form by instantiating them '''
+        self.forms = [i(*args) for i in self.form_class]
```
```suggestion
    @classmethod
    def create_forms(cls, *args):#request_method='GET'
        ''' Set default values for the form by instantiating them '''
        return [i(*args) for i in cls.form_class]
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,91 @@
+
+    def get(self, request, *args, **kwargs):
+        ''' Render page if no data is given '''
+        self.create_forms()
```
```suggestion
        forms = self.create_forms()
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,91 @@
+        return CatalogDataset.objects.all()
+
+
+    def validate_forms(self):
+        for element_forms in self.forms:  # for loop for making clean data by is_valid() method
+            element_forms.is_valid()
+            for errorField in element_forms.errors:  # temporary error message
+                print(f"errorField")
```
```suggestion
    @classmethod
    def validate_forms(self, forms):
        for form in forms:  # for loop for making clean data by is_valid() method
            form.is_valid()
        return forms
```

aanersc commented:
@akorosov @aperrin66 
The tests are added in 85a49de. I have removed the tests of old viewer.
These tests includes integration tests in addtion to unittests.


There is a new migration which travis needed for testing. This file is correctly pushed in the 7a1427e.

There a minor changes in runtests.py which is corrected in c9c5afc. I have also added a separate static file based on recommendation of @aperrin66 .

Please have a look at them carefully.

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,74 @@
+
+#from geospaas.base_viewer.models import Search_with_para
+
+class BaseForm(forms.ModelForm):
```
I am having trouble to understand why you use `ModelForm` as the base class.
From what I understand, `ModelForm` is used when you want to create/update instances of Models in your database.
You only have search forms here, so you can probably just use the `Form` class as base.

aanersc replied:
django `ModelForm` is a 'direct bridge' of our models. These kind of forms are specified based on the specifications of columns of our database. Thus they are easily set and used and the user can correctly behave with them since they are individually customized with individuals of our columns of database. This is an ability of Django which make things easy for us to use this kind of form and then stack them in a 'html form' for further processing. In this way, we can do what we want in an `form agnostic` way in `views.py`. `form agnostic` by our mean is that `do the logic in views regardless of type of form and type of data that we work with it`. This might not possible without django `ModelForm`(or maybe extreme difficult and needs lots of lines of code to be operational).

aperrin66 replied:
Actually, you don't use much of the functionalities of `ModelForm`, so I am pretty sure you can just replace your forms by the following:
```
class TimeAndSourceForm(forms.Form):
    time_coverage_start = forms.DateTimeField(initial=timezone.datetime(2000, 1, 1))
    time_coverage_end = forms.DateTimeField(initial=timezone.now())
    source = forms.ModelChoiceField(Source.objects.all())

    def filter(self, ds):
        [...]

class SpatialSearchForm(forms.Form):
    polygon = forms.GeometryField(widget=LeafletWidget(attrs={
        'settings_overrides': {
            'DEFAULT_CENTER': (60.0, 5.0),
            'DEFAULT_ZOOM': 1,
            'PLUGINS': {'forms': {'auto-include': True}},
        }
    }))

    def filter(self, ds):
        [...]
```

This also removes the need for the `BaseForm` class and the `SearchModelForLeaflet` model.

aanersc replied:
YES, you have also a corrcet approach. But when django automate things for us, why don't we take advantage of it? All of those `DateTimeField`, and`ModelChoiceField` explicit definition of fields can be done by django implicitly(automatically). This way we can inherit from this form in the ADASviewer and remove each of these fields when we want them to remove. Modelform provides more flexibility for us to **remove** or **add** fields in possible inheritance in future.

I discussed it previously with @akorosov. I had your opinion previously,but now I have found that this could be the better way of implementing it.

aperrin66 commented on geospaas/base_viewer/models.py:
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+from django.contrib.gis.db import models as geomodels
+
+class SearchModelForLeaflet(geomodels.Model):
```
Why do you create this extra model? Is it necessary to save searches in the database?

aanersc replied:
leaflet needs this model to work with its widget. This information comes thoroughly from the user and it is not related to our model and database. Thus, it is good to use its specified model as described in the tutorial of leaflet.

aperrin66 replied:
Maybe I am mistaken but I don't think it is really necessary, see my last comment on the forms.

akorosov commented on geospaas/base_viewer/tests.py:
```diff
@@ -0,0 +1,145 @@
+class UnitTestingForViews(TestCase):
+    fixtures = ["vocabularies", "catalog"]
+
+    def setUp(self):
+        self.v1 = IndexView()
+        self.ds = self.v1.get_all_datasets()
+        self.forms = self.v1.create_forms(self)
+
+    def test_creating_the_forms(self):
+        '''shall create the specified fields of forms'''
+        self.assertTrue('polygon' in self.forms.__str__())
+        self.assertTrue('time_coverage_start' in self.forms.__str__())
+        self.assertTrue('time_coverage_end' in self.forms.__str__())
+        self.assertTrue('source' in self.forms.__str__())
```
Since we switch to Forms the examples above are not valid anymore. An example on how to write a unit test for IndexView. A valid test example is here:
```python
class IndexViewTests(TestCase):
    @patch('geospaas.base_viewer.views.CatalogDataset')
    def test_get_all_datasets(self, mock_dataset):
        """ Shall call CatalogDataset.objects.all() inside get_all_datasets """
        ds = IndexView.get_all_datasets()
        mock_dataset.objects.all.assert_called_once()
```
Please pay attention to the name of the class, name of the method, dockstring. These are quite standard. Don't invent anything else.

aanersc commented:
@aperrin66 @akorosov 
In the d9888bc, the models.py is complete deleted and the for loops are also removed from views.py. There are some tests added to be the unittest. please check them carefully.

akorosov commented on geospaas/base_viewer/forms.py:
```diff
@@ -1,74 +1,43 @@
 
-class BaseForm(forms.ModelForm):
-    initial = dict()
+class OverallForm(forms.Form):
```
Can you call it BaseForm? 
```suggestion
class BaseForm(forms.Form):
```

akorosov commented on geospaas/base_viewer/templates/base_viewer/template_for_base.html:
```diff
@@ -54,9 +54,11 @@
 
-    {{TimeAndSourceForm.as_p}}
+
+    {{OverallForm.as_p}}
```
It is better not to hardcode here a name of the form class. Because when we override IndexView we will have to change the name of the form class. Then we would also have to change the template. But if in opposite we keep a generic name (say 'form') here (and, hence, in set_context) then we are able to reuse this template.
```suggestion
    {{form.as_p}}
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -1,73 +1,51 @@
-                if (isinstance(form.__class__(), j )):
-                    context[j.__name__] = form
-
+        context[cls.form_class.__name__] = form
```
It is better not to hardcode here a name of the form class. Otherwise we have to hardcode the name of the form class in template and we will not be able to reuse the template.
```suggestion
        context['form'] = form
```

akorosov commented on geospaas/base_viewer/forms.py:
```diff
@@ -1,74 +1,43 @@
 
-from geospaas.base_viewer.models import SearchModelForLeaflet
-from geospaas.catalog.models import Dataset as CatalogDataset
+from geospaas.catalog.models import Source as CatalogSource
```
Since you don't define new Source in this module, you don't have to import it with another name.
```suggestion
from geospaas.catalog.models import Source
```

akorosov commented on geospaas/base_viewer/forms.py:
```diff
@@ -1,74 +1,43 @@
+        initial=timezone.datetime(2000, 1, 1))
+    time_coverage_end = forms.DateTimeField(initial=timezone.now())
+    source = forms.ModelChoiceField(
+        CatalogSource.objects.all(), required=False)
```
```suggestion
        Source.objects.all(), required=False)
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -1,73 +1,51 @@
 
-from geospaas.base_viewer.forms import TimeAndSourceForm, SpatialSearchForm
+from geospaas.base_viewer.forms import OverallForm
 from geospaas.catalog.models import Dataset as CatalogDataset
```
Since you don;t define a new Dataset in this module, you don't have to import it with another name
```suggestion
from geospaas.catalog.models import Dataset
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -1,73 +1,51 @@
 
     @classmethod
     def get_all_datasets(cls):
         return CatalogDataset.objects.all()
```
```suggestion
        return Dataset.objects.all()
```

akorosov commented on geospaas/base_viewer/tests.py:
```diff
@@ -102,44 +103,25 @@ def test_the_get_verb_of_GUI(self):
+        form = MagicMock()
+        ds = MagicMock()
+        context = IndexView.set_context(form,ds)
+        form.filter.assert_not_called()
```
Not really needed here.
```suggestion
```

aanersc replied:
I think it is better to have it here because this line makes sure that all datasets are shown and no one is filtered. If in the future development the filter function is called incorrectly, this line prevents it. It is needed to show a **_comprehensive_** result. Kindly, how do you think? Is it still not needed?

akorosov replied:
OK, it is a minor thing anyway

aperrin66 commented on geospaas/base_viewer/templates/base_viewer/origin_of_templates.html:
```diff
@@ -0,0 +1,33 @@
+{% load static %}
```
About the names of the templates files:
  - you probably don't need to include the word "template" in the names, since they are in the `templates` directory
  - in would be nice to have names that describe what the templates contain, for example `page_structure.html` and `elements.html`

aanersc replied:
done

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,43 @@
+        t0 = self.cleaned_data['time_coverage_start']
+        t1 = self.cleaned_data['time_coverage_end'] + \
+            timezone.timedelta(hours=24)
+        ds = ds.filter(Q(time_coverage_end__lt=t1) &
+                       Q(time_coverage_start__gt=t0))
```
thanks to your recommendation, it is improved.

aanersc commented on geospaas/base_viewer/tests.py:
```diff
@@ -0,0 +1,127 @@
+
+
+class IntegrationTestsForGUIWithNewBase(TestCase):
+    '''INtegration tests for GET and POST methods of GUI'''
```
done

aanersc commented on geospaas/base_viewer/tests.py:
```diff
@@ -0,0 +1,127 @@
+from geospaas.base_viewer.views import IndexView
+
+
+class MyHTMLParser(HTMLParser):
```
done.

aanersc commented on geospaas/viewer/urls.py:
```diff
@@ -4,5 +4,5 @@
 app_name = 'viewer'
 urlpatterns = [
-    url(r'^$', IndexView.as_view(), name='index'),
+    url('', IndexView.as_view(), name='index'),
```
when there is no need to use regex, then it is useless to use it. Django automatically finds its way! It is simpler, this is the only reason(as far as I know).

aanersc commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,43 @@
+                                      }
+                                  }))
+    time_coverage_start = forms.DateTimeField(
+        initial=timezone.datetime(2000, 1, 1))
```
thanks to your comment, I try to fix it. please kindly attention to my new code for this topic.

aanersc commented on geospaas/base_viewer/tests.py:
```diff
@@ -0,0 +1,127 @@
+        """shall return only the first dataset of fixtures in the specified placement of datasets inside the resulted HTML
+        in the case of a POST request with a good choice of polygon"""
+        res1 = self.client.post('/', {'polygon': '{"type":"Polygon","coordinates":[[[0,0],[0,5],[5,5],[5,0],[0,0]]]}',
+                                      'time_coverage_start': timezone.datetime(2000, 12, 29),
+                                      'time_coverage_end': timezone.datetime(2020, 1, 1),
```
It is not a very necessary consideration in the tests. Time lag between different places does not matter for tests.

aperrin66 replied:
First of all, it seems pretty important that we test that datasets are correctly filtered based on dates.
Second, if you mix aware and naive datetimes, you will end up getting errors because you can't compare them.

aanersc replied:
OK. Then let me see how precisely the previous tests (which are not written by me) are accomplishing regarding the time in the previous version of viewer and also check that if tzinfo matters for them or not. If you say it is important, then it is important.  :)

aanersc commented on geospaas/base_viewer/tests.py:
```diff
@@ -0,0 +1,127 @@
+        self.assertEqual(res1.status_code, 200)
+        self.parser.feed(str(res1.content))
+        # the first dataset of fixtures must be in the html
+        self.assertIn(True, [('file://localhost/some/test/file1.ext' in dat)
+                             for dat in self.parser.data])
```
done

aperrin66 replied:
It was just an information, the way you did it was perfectly valid.

aanersc commented on geospaas/viewer/tests.py:
```diff
@@ -1,137 +0,0 @@
-import os
```
Yeah, I am waiting for your feedbacks about this new one to remove the old one at the end of the branch.

aanersc commented on geospaas/base_viewer/templates/base_viewer/template_for_base.html:
```diff
@@ -0,0 +1,69 @@
+{% block searchbox %}
+
+{% endblock searchbox %}
+##############################################################################################################
```
:)

aanersc commented on geospaas/base_viewer/templates/base_viewer/template_for_base.html:
```diff
@@ -0,0 +1,69 @@
+{% extends "base_viewer/origin_of_templates.html" %}
+
+{% load leaflet_tags %}
+
+{% leaflet_js %}
+{% leaflet_css %}
+
+{% block searchbox %}
+
+{% endblock searchbox %}
```
removed

aanersc commented on geospaas/base_viewer/migrations/0001_initial.py:
```diff
@@ -0,0 +1,22 @@
+# Generated by Django 3.0 on 2020-05-22 13:46
+
+import django.contrib.gis.db.models.fields
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    initial = True
+
+    dependencies = [
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='SearchModelForLeaflet',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('polygon', django.contrib.gis.db.models.fields.PolygonField(blank=True, null=True, srid=4326)),
+            ],
+        ),
+    ]
```
done

aanersc commented:
@akorosov @aperrin66 
Minor comments which are discussed in github are addressed in 229bcb1. There is a tiny modification in 229bcb1 which is reverted into original state in 9b8dac2. Please check the 229bcb1 as the feedback of your comments. The old version of viewer is still in the branch. The old version will remove based on the comment of @aperrin66 by alongside introducing more tests in the future commit.

akorosov commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,42 @@
+    source = forms.ModelChoiceField(
+        Source.objects.all(), required=False)
+
+    def filter(self, ds):
```
Add dockstring.

aanersc replied:
done

akorosov commented on geospaas/base_viewer/templates/base_viewer/page_structure.html:
```diff
@@ -0,0 +1,33 @@
+			{% endblock content_result%}
+		</div><!--end of result section-->
+
+		<div id="searchbox"><!--start of searchbox-->
+			{% block searchbox %}
+			{% endblock searchbox %}
+		</div><!--end of searchbox-->
```
Where is `searchbox` defined? I could find only `content_result` and `contentmap` in `elements.html`

aanersc replied:
It is not used at all. Maybe used for ADAS viewer. Right now, it is just an empty tag of HTML

akorosov replied:
Can you remove it? If we need it in ADAS, it is better to extend page_structure.html and add something than have an empty, potentially-non needed placeholder.

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,49 @@
+    viewname = 'index'
+
+    @classmethod
+    def get_all_datasets(cls):
```
Docstring is missing

aanersc replied:
done

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,49 @@
+        return form.filter(ds)
+
+    @classmethod
+    def set_context(cls, form, ds):
```
Docstring is missing

aanersc replied:
done

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,49 @@
+        return context
+
+    @classmethod
+    def set_dataset_context(cls, ds):
```
Docstring is missing

aanersc replied:
done

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,49 @@
+        return ds  # excluded from other context for pagination development of django
+
+    def get(self, request, *args, **kwargs):
+        ''' Render page if no data is given '''
```
```suggestion
        """ Render page if no data is given """
```

akorosov commented on geospaas/base_viewer/views.py:
```diff
@@ -0,0 +1,49 @@
+        return render(request, self.main_template, context)
+
+    def post(self, request, *args, **kwargs):
+        ''' all sections needed for POST requests '''
```
```suggestion
        """ Render page when user submits search request """
```

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,42 @@
+from django.contrib.gis import forms
+from django.utils import timezone
+from datetime import tzinfo
```
Unused import

aanersc replied:
done

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,42 @@
+from django.utils import timezone
+from datetime import tzinfo
+from leaflet.forms.widgets import LeafletWidget
+import pytz
```
You don't need that, since you already import `timezone` from `django.utils`, you can use `timezone.utc` instead of `pytz.utc`.

aanersc replied:
done

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,42 @@
+                                          'PLUGINS': {'forms': {'auto-include': True}},
+                                      }
+                                  }))
+    utc = pytz.utc
```
This should not be a class attribute, can you directly use `timezone.utc` as the `tzinfo` argument in `time_coverage_start`?

aanersc replied:
Yes, done.

aperrin66 commented on geospaas/base_viewer/forms.py:
```diff
@@ -0,0 +1,42 @@
+    utc = pytz.utc
+    time_coverage_start = forms.DateTimeField(
+        initial=timezone.datetime(2000, 1, 1, tzinfo=utc))
+    time_coverage_end = forms.DateTimeField(initial=timezone.now().utcnow())
```
`timezone.now().utcnow()` actually returns the current UTC time as a naive datetime, so it won't do here.
(Plus you call the class method `utcnow()` from the `datetime` class after creating a datetime instance with `now()`, which is inefficient.)
See https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow

You can either:
  - keep `timezone.now()`, which will return an aware datetime if the Django `USE_TZ` setting is set to `True`.
  - if you want to be sure to always get an aware datetime, you can use the `now()` method of the `datetime` module. That would look like this:
```
from datetime import datetime, timezone
[...]
time_coverage_end = forms.DateTimeField(initial=datetime.now(tz=timezone.utc))
[...]
```

aanersc replied:
Thanks. I prefer the first choice. In complementary note to your explanation, as shown in the definition of `now` in Django, using `now` is faster without argument:
![image](https://user-images.githubusercontent.com/62345618/83630774-acc08200-a59c-11ea-8f7c-11a9b0df29bb.png)
So the first choice seems better!

aperrin66 replied:
OK, but there might be a misunderstanding about the comment in the code: these two ways of creating an aware `datetime` object containing the current UTC time yield the same result:
- `datetime.now(tz=utc)`
- `datetime.utcnow().replace(tzinfo=utc)`

The second one is apparently faster, which is why it is used in `django.utils.timezone.now()`.

But be careful not to confuse `django.utils.timezone.now()` and `datetime.datetime.now()` methods.
When you call `django.utils.timezone.now()` without argument, it will call `datetime.utcnow().replace(tzinfo=utc)` if `USE_TZ` is true (which should be the case for us).
Otherwise, it will call `datetime.datetime.now()` and return a naive `datetime` object.

akorosov commented on tests/settings.py:
```diff
@@ -31,6 +31,7 @@
     'leaflet',
     'django_forms_bootstrap',
     'geospaas.viewer',
```
```suggestion
```

aanersc commented:
Ready to merge for 9e5cb49
@aperrin66 @akorosov
